### ISSUE TITLE
[🚀 Feature]: if session fails >> choose other available node till 5-10 times

### ISSUE THREAD

**Issue by krishtoautomate** [SeleniumHQ/selenium#13358]
### Feature and motivation

sometimes session fails to create and leads to test fail in configuration. so if feature is added with on session fails due to errors from node, the create session should retry on different available node with similar desired capabilities.

### Usage example

for session or new driver instance

**Comment by pujagani** [comment]
What would the reason be for not being able to create the session that requires a retry that already does not exist?
I think have covered most situations that would require a retry. 
If the slots are full when the session is created, then the session is retried. We have that mechanism in place. If the slot does not match the capability for some reason, then it is not retried.

**Comment by krishtoautomate** [comment]
It can be any reason - node is up but have issues with connection or driver, driver version match or socket errors etc which leads to session fail

**Comment by pujagani** [comment]
A WebSocket connection(for CDP or BiDi) is made after the session is created, so retrying an already-created session will not be valid at that point. Websocket connection creation errors are mainly if the underlying system (example a proxy server in between) does not allow creating a WebSocket connection or if the WebSocket connection is closed from the user end. Neither are valid candidates for retry. 
Browser version match is handled as part of matching the slots. 
Driver version mismatch might happen when using CDP, but retrying might not make sense since the driver version will not update automatically within a few seconds of retry attempt. That needs fixing from the user's end and it is not classified as intermittent issue. 

If the Node is up and slot is free, and it is unable to create a session due to a driver error, we need to know the exact error to  retry since on session creation if HTTP status code in response is 200 then we are good else the webdriver returns error code 500 (https://www.w3.org/TR/webdriver2/#new-session), which is not sufficient to know if the error is intermittent or not to retry.

**Comment by krishtoautomate** [comment]
my intention is not retry to same node if driver mismatch error or any other error occurs on node, as you said - nodes will not get fixed by its own without fix. the slots should reshuffle and error occurred nodes should be pushed to last slots and  session attempt should retry in new nodes.

**Comment by titusfortner** [comment]
This needs to be managed in client code; if you think it is a recoverable error, catch it and retry. Anything done on the Selenium side would need to make assumptions about the state of the system that may or may not be true.

**Comment by krishtoautomate** [comment]
Retry can be managed on client but if slot is not shuffled, it ll try to connect to same node and trying will not make sense.

**Comment by titusfortner** [comment]
@diemol does grid randomize node selection or is it in a queue?

If a node is bad and won't be fixed, I think client needs to shut it down, not have an automatic feature to avoid it: https://www.selenium.dev/documentation/grid/advanced_features/endpoints/#drain

**Comment by diemol** [comment]
>@diemol does grid randomize node selection or is it in a queue?

Mainly, it orders Nodes by load.

Retrying by default does not make sense because it hides problems in Nodes; if a Node has a wrong configuration, it must be removed from the Grid. Retrying can eventually help you once, but in the long term, you will have a slow Grid because you somehow end up using fewer Nodes than the ones you have up and running. In addition, debugging this issue would be very hard since you would perceive slowness and not know why. 

The Slot selection is made by the [DefaultSlotSelector](https://github.com/SeleniumHQ/selenium/blob/48b5659afd7c7b797a07641a07c0ad223e69ed39/java/src/org/openqa/selenium/grid/distributor/selector/DefaultSlotSelector.java#L32). The good news is that you can implement your own `SlotSelector` and override the default one by passing the `--slot-selector` flag to the Distributor. 

In short, this is already possible, and you can do it. If you need help with it, feel free to drop by our chat: https://www.selenium.dev/support/#ChatRoom

**Comment by krishtoautomate** [comment]
this piece of code will make reuse same faulty node again and again till all tests fail:

  // Then last session created (oldest first), so natural ordering again
                .thenComparingLong(NodeStatus::getLastSessionCreated)

is their anyway even tried node to pushed to newest list
